File: backend/app/domains/roles/schemas/roles.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.models.enums import OrganizationRole
from app.models.permissions import PERMISSION_CATALOG, ROLE_PERMISSIONS

_ALL_VALID_PERMISSIONS = {entry["permission"] for entry in PERMISSION_CATALOG}
_ALL_ROLE_VALUES = {r.value for r in OrganizationRole}


class PermissionEntry(BaseModel):
    permission: str
    category: str
    description: str


class PermissionCatalogResponse(BaseModel):
    items: list[PermissionEntry]
    total: int


class BuiltinRoleResponse(BaseModel):
    role: str
    label: str
    description: str
    permissions: list[str]
    is_builtin: bool = True


class CustomRoleResponse(BaseModel):
    id: str
    organization_id: str
    name: str
    description: str | None
    base_role: str | None
    permissions: list[str]
    created_by_id: str | None
    created_at: datetime
    updated_at: datetime
    is_builtin: bool = False


class RoleListResponse(BaseModel):
    builtin_roles: list[BuiltinRoleResponse]
    custom_roles: list[CustomRoleResponse]


class CreateCustomRoleRequest(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    description: str | None = Field(default=None, max_length=512)
    base_role: str | None = None
    permissions: list[str] = Field(default_factory=list)

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("name must not be blank")
        if stripped.lower() in _ALL_ROLE_VALUES:
            raise ValueError("name conflicts with a built-in role name")
        return stripped

    @field_validator("base_role")
    @classmethod
    def validate_base_role(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if value not in _ALL_ROLE_VALUES:
            raise ValueError(f"base_role must be one of {sorted(_ALL_ROLE_VALUES)}")
        return value

    @field_validator("permissions")
    @classmethod
    def validate_permissions(cls, values: list[str]) -> list[str]:
        unknown = [p for p in values if p not in _ALL_VALID_PERMISSIONS]
        if unknown:
            raise ValueError(f"unknown permissions: {unknown}")
        return list(dict.fromkeys(values))  # deduplicate, preserve order


class UpdateCustomRoleRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=64)
    description: str | None = None
    base_role: str | None = None
    permissions: list[str] | None = None

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        if not stripped:
            raise ValueError("name must not be blank")
        if stripped.lower() in _ALL_ROLE_VALUES:
            raise ValueError("name conflicts with a built-in role name")
        return stripped

    @field_validator("base_role")
    @classmethod
    def validate_base_role(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if value not in _ALL_ROLE_VALUES:
            raise ValueError(f"base_role must be one of {sorted(_ALL_ROLE_VALUES)}")
        return value

    @field_validator("permissions")
    @classmethod
    def validate_permissions(cls, values: list[str] | None) -> list[str] | None:
        if values is None:
            return None
        unknown = [p for p in values if p not in _ALL_VALID_PERMISSIONS]
        if unknown:
            raise ValueError(f"unknown permissions: {unknown}")
        return list(dict.fromkeys(values))


# Descriptions shown in the UI for each built-in role.
BUILTIN_ROLE_META: dict[str, dict[str, str]] = {
    "owner": {
        "label": "Owner",
        "description": "Full access to all features including billing and role management. Cannot be changed.",
    },
    "admin": {
        "label": "Admin",
        "description": "Full access except billing. Can manage team, roles, security, and all content.",
    },
    "member": {
        "label": "Member",
        "description": "Can upload documents, use chat, and view collections.",
    },
    "viewer": {
        "label": "Viewer",
        "description": "Read-only access to documents, collections, and chat.",
    },
    "reviewer": {
        "label": "Reviewer",
        "description": "Can create and run evaluations and view audit logs in addition to viewer access.",
    },
    "developer": {
        "label": "Developer",
        "description": "Can manage API keys, webhooks, and create agents in addition to member access.",
    },
    "security_admin": {
        "label": "Security Admin",
        "description": "Dedicated security role: can configure SSO/SCIM, view audit logs, and manage security policies.",
    },
    "billing_admin": {
        "label": "Billing Admin",
        "description": "Dedicated billing role: can view and manage billing without access to content.",
    },
}


def builtin_roles_response() -> list[BuiltinRoleResponse]:
    result = []
    for role in OrganizationRole:
        meta = BUILTIN_ROLE_META.get(role.value, {"label": role.value.title(), "description": ""})
        perms = sorted(ROLE_PERMISSIONS.get(role.value, frozenset()))
        result.append(
            BuiltinRoleResponse(
                role=role.value,
                label=meta["label"],
                description=meta["description"],
                permissions=perms,
            )
        )
    return result

File: frontend/src/lib/api/roles.ts
import { apiRequest } from "@/lib/api/request";

export type PermissionEntry = {
  permission: string;
  category: string;
  description: string;
};

export type PermissionCatalogResponse = {
  items: PermissionEntry[];
  total: number;
};

export type BuiltinRole = {
  role: string;
  label: string;
  description: string;
  permissions: string[];
  is_builtin: true;
};

export type CustomRole = {
  id: string;
  organization_id: string;
  name: string;
  description: string | null;
  base_role: string | null;
  permissions: string[];
  created_by_id: string | null;
  created_at: string;
  updated_at: string;
  is_builtin: false;
};

export type RoleListResponse = {
  builtin_roles: BuiltinRole[];
  custom_roles: CustomRole[];
};

export type CreateCustomRoleRequest = {
  name: string;
  description?: string | null;
  base_role?: string | null;
  permissions: string[];
};

export type UpdateCustomRoleRequest = {
  name?: string | null;
  description?: string | null;
  base_role?: string | null;
  permissions?: string[] | null;
};

const ROLES_BASE = "/admin/roles";

function normalizePermissionEntry(value: unknown): PermissionEntry {
  const raw =
    value && typeof value === "object"
      ? (value as Record<string, unknown>)
      : {};
  return {
    permission: typeof raw.permission === "string" ? raw.permission : "",
    category: typeof raw.category === "string" ? raw.category : "",
    description: typeof raw.description === "string" ? raw.description : "",
  };
}

function normalizeBuiltinRole(value: unknown): BuiltinRole {
  const raw =
    value && typeof value === "object"
      ? (value as Record<string, unknown>)
      : {};
  return {
    role: typeof raw.role === "string" ? raw.role : "",
    label: typeof raw.label === "string" ? raw.label : "",
    description: typeof raw.description === "string" ? raw.description : "",
    permissions: Array.isArray(raw.permissions)
      ? (raw.permissions as unknown[]).filter(
          (p): p is string => typeof p === "string",
        )
      : [],
    is_builtin: true,
  };
}

function normalizeCustomRole(value: unknown): CustomRole {
  const raw =
    value && typeof value === "object"
      ? (value as Record<string, unknown>)
      : {};
  return {
    id: typeof raw.id === "string" ? raw.id : "",
    organization_id:
      typeof raw.organization_id === "string" ? raw.organization_id : "",
    name: typeof raw.name === "string" ? raw.name : "",
    description: typeof raw.description === "string" ? raw.description : null,
    base_role: typeof raw.base_role === "string" ? raw.base_role : null,
    permissions: Array.isArray(raw.permissions)
      ? (raw.permissions as unknown[]).filter(
          (p): p is string => typeof p === "string",
        )
      : [],
    created_by_id:
      typeof raw.created_by_id === "string" ? raw.created_by_id : null,
    created_at: typeof raw.created_at === "string" ? raw.created_at : "",
    updated_at: typeof raw.updated_at === "string" ? raw.updated_at : "",
    is_builtin: false,
  };
}

export async function listPermissions(): Promise<PermissionCatalogResponse> {
  const payload = await apiRequest<unknown>(`${ROLES_BASE}/permissions`, {
    method: "GET",
    retry: false,
  });
  const raw =
    payload && typeof payload === "object"
      ? (payload as Record<string, unknown>)
      : {};
  const items = Array.isArray(raw.items)
    ? raw.items.map(normalizePermissionEntry)
    : [];
  return {
    items,
    total: typeof raw.total === "number" ? raw.total : items.length,
  };
}

export async function listRoles(): Promise<RoleListResponse> {
  const payload = await apiRequest<unknown>(ROLES_BASE, {
    method: "GET",
    retry: false,
  });
  const raw =
    payload && typeof payload === "object"
      ? (payload as Record<string, unknown>)
      : {};
  return {
    builtin_roles: Array.isArray(raw.builtin_roles)
      ? raw.builtin_roles.map(normalizeBuiltinRole)
      : [],
    custom_roles: Array.isArray(raw.custom_roles)
      ? raw.custom_roles.map(normalizeCustomRole)
      : [],
  };
}

export async function getCustomRole(roleId: string): Promise<CustomRole> {
  const payload = await apiRequest<unknown>(
    `${ROLES_BASE}/${encodeURIComponent(roleId)}`,
    { method: "GET", retry: false },
  );
  return normalizeCustomRole(payload);
}

export async function createCustomRole(
  request: CreateCustomRoleRequest,
): Promise<CustomRole> {
  const payload = await apiRequest<unknown>(ROLES_BASE, {
    method: "POST",
    json: {
      name: request.name,
      description: request.description ?? null,
      base_role: request.base_role ?? null,
      permissions: request.permissions,
    },
    retry: false,
  });
  return normalizeCustomRole(payload);
}

export async function updateCustomRole(
  roleId: string,
  request: UpdateCustomRoleRequest,
): Promise<CustomRole> {
  const payload = await apiRequest<unknown>(
    `${ROLES_BASE}/${encodeURIComponent(roleId)}`,
    { method: "PATCH", json: request, retry: false },
  );
  return normalizeCustomRole(payload);
}

export async function deleteCustomRole(roleId: string): Promise<void> {
  await apiRequest<unknown>(`${ROLES_BASE}/${encodeURIComponent(roleId)}`, {
    method: "DELETE",
    retry: false,
  });
}
